plugin/pkg/scheduler/scheduler.go

```diff
@@ -265,6 +304,7 @@ func (sched *Scheduler) scheduleOne() {
 	suggestedHost, err := sched.schedule(pod)
 	metrics.SchedulingAlgorithmLatency.Observe(metrics.SinceInMicroseconds(start))
 	if err != nil {
+		sched.preempt(pod, err)
```

Comment:
I added checks to `selectNodesForPreemption` to check extenders. The internal scheduler find the nodes and the victims first and then sends those nodes for further filtering to the extenders. I added a test case to verify the logic.
This addresses scenario (2), but scenario (1) will still cause some wasted work if a pod is determined to be infeasible by extenders only. That would require some more refactoring to distinguish between failed predicates reported by the internal scheduler and the extenders. I guess we can safely leave that for a future optimization.
How do you like the new logic?

Reply:
So IIUC the new logic is
(1) using the internal scheduler logic, find set of nodes that are feasible with preemption, and the minimal set of victims for each node
(2) for each node, ask the scheduler extender(s) whether the pod is feasible on the node with the victims gone. For nodes where the answer is "no," remove from the set of feasible nodes.

I think that works. IIUC the wasted work you are referring to is computing victims using the internal scheduler, when you will be infeasible anyway due to the extenders? (In other words you're saying that if schedule() could tell you it is infeasible due to the extender, you could immediately remove the node and not bother with that node in selectNodesForPreemption() at all?) It seems like that happens in both scenarios since preemption will never fix infeasibility due to the extender -- or am I misunderstanding?

BTW I'm not sue the extender will always see the preempted version of the node; I don't remember the details but you should trace through what extender.Filter() actually ends up sending the the extender (there is some kind of "NodeCacheCapable" mode that affects things, but I don't remember how it works). This shouldn't affect correctness, I am just mentioning it. It shouldn't matter whether it gets the unpreempted version of the node or the preempted version of the node, but it would be good to make sure you understand that codepath anyway.

Reply:
Yes, your understanding it correct. The algorithm is exactly as you described and the wasted work is exactly as you mentioned. We wouldn't bother with a node if we knew that the pod was determined to be infeasible on the node by the extenders only. This happens in both scenarios, but in the second scenario, removal of some pods that make the pending pod feasible by the internal scheduler, might also make the pod feasible by the extenders. So, there is a probability that the work is not completely wasted, but in scenario 1, no pod will be removed from the node and all the checks will be wasted.

I will check NodeCacheCapable.

Reply:
I checked NodeCacheCapable. Looks like the only difference is that when NodeCacheCapable is true, we only send node names to the extender instead of sending node objects. So, it looks to me that the current implementation should work for both cases.